File: linked-lists/merge-two-sorted-lists/merge-two-sorted-lists.js
class ListNode {
  val;
  next;

  constructor(val, next) {
    this.val = val;
    this.next = next;
  }
}

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
function mergeTwoLists(l1, l2) {
  let i = l1;
  let j = l2;

  const dummyNode = new ListNode();
  let currentNode = dummyNode;

  while (i !== null && j !== null) {
    const iVal = i.val;
    const jVal = j.val;

    if (iVal <= jVal) {
      currentNode.next = i;
      currentNode = currentNode.next;
      i = i.next;
    } else {
      currentNode.next = j;
      currentNode = currentNode.next;
      j = j.next;
    }
  }

  if (i !== null) {
    currentNode.next = i;
  } else {
    currentNode.next = j;
  }

  return dummyNode.next;
}

File: linked-lists/merge-two-sorted-lists/merge-two-sorted-lists.go
package mergetwosortedlists

// ListNode is a node for a Linked List
type ListNode struct {
	Val  int
	Next *ListNode
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	i := l1
	j := l2

	dummyNode := &ListNode{}
	currentNode := dummyNode

	for i != nil && j != nil {
		iVal := i.Val
		jVal := j.Val

		if iVal <= jVal {
			currentNode.Next = i
			currentNode = currentNode.Next
			i = i.Next
		} else {
			currentNode.Next = j
			currentNode = currentNode.Next
			j = j.Next
		}
	}

	if i != nil {
		currentNode.Next = i
	} else {
		currentNode.Next = j
	}

	return dummyNode.Next
}
